### PR TITLE
feat(ProgressIndicator): add optional interactivity

### PR DESCRIPTION
**Pull request by scottdickerson**
Closes IBM/carbon-components-react#1944

{{short description}}

#### Changelog

**New**

- Added new onChange prop to ProgressIndicator that will attach a keyboard/click handler to each ProgressStep. It will callback with the selected index.  This should be accessible and clickable.

### CODE REVIEW

File: src/components/ProgressIndicator/ProgressIndicator.js [hunk 1]
```diff
@@ -37,6 +38,14 @@ export const ProgressStep = ({ ...props }) => {
     [className]: className,
   });
 
+  const handleKeyDown = e => {
+    const key = e.key || e.which;
+
+    if (key === 'Enter' || key === 13 || key === ' ' || key === 32) {
```
Comment (emyarod) [inline src/components/ProgressIndicator/ProgressIndicator.js]: we may be able to use our [keyboard event handler helpers](https://github.com/IBM/carbon-components-react/blob/master/src/tools/key.js) here

File: src/components/ProgressIndicator/ProgressIndicator.js [hunk 2]
```diff
@@ -88,21 +97,37 @@ export const ProgressStep = ({ ...props }) => {
...
-      <ProgressStepLabel className={`${prefix}--progress-label`}>
-        {label}
-      </ProgressStepLabel>
-      {componentsX &&
-      secondaryLabel !== null &&
-      secondaryLabel !== undefined ? (
-        <p className={`${prefix}--progress-optional`}>{secondaryLabel}</p>
-      ) : null}
-      <span className={`${prefix}--progress-line`} />
+      <div
```
Comment (emyarod) [inline src/components/ProgressIndicator/ProgressIndicator.js]: it appears that the experimental version of this component is broken with this change. I think we may have to incorporate the new markup and add new style rules to the vanilla component before we can merge this into the React repo

**Comment by emyarod** [comment]
thanks for jumping in @scottdickerson! I think functionally this PR is solid, but it looks like these changes will break the appearance of the experimental version, so we will probably first implement these changes on the vanilla side before merging this PR in

**Comment by scottdickerson** [comment]
thank you for catching these issues @emyarod I've updated the PR and I think fixed both, can you take another look?

**Comment by emyarod** [comment]
thanks for the quick update @scottdickerson! It looks like overflowing content is not showing the ellipsis and most likely requires tweaks from the vanilla side if I'm not mistaken, which I'm happy to help out with once we get a ticket opened up for style and markup updates over there

image

image

I also wonder if the design team wants to modify the focus state so that it's not the browser default 
cc @IBM/carbon-designers 

image

image

File: src/components/ProgressIndicator/ProgressIndicator.js [hunk 3]
```diff
@@ -88,21 +96,37 @@ export const ProgressStep = ({ ...props }) => {
...
+        style={{
+          display: 'inline-flex',
```
Comment (asudoh) [inline src/components/ProgressIndicator/ProgressIndicator.js]: Wondering what this style is for, but if this style is absolutely necessary, could you please create a PR to https://github.com/IBM/carbon-components/pulls with the corresponding style ruleset?

**Comment by joshblack** [comment]
Hey @scottdickerson! 👋 Just wanted to check in on this PR, would you have time to address @asudoh's comment over in: https://github.com/IBM/carbon-components-react/pull/1945#discussion_r261785240 ?

**Comment by scottdickerson** [comment]
sure @joshblack I'll adjust the styles under this PR for the vanilla comp

https://github.com/IBM/carbon-components/issues/1652

**Comment by scottdickerson** [comment]
ok added the requested changes, please take a look @asudoh and @emyarod when you get a chance.  If the carbon-components PR is accepted and merged, please rev the dependency to match to pickup the new styles.

**Comment by scottdickerson** [comment]
@joshblack I updated the dep to point to the latest carbon-components that has the new styles, can you take a look?  Thanks!

File: src/components/ProgressIndicator/ProgressIndicator.js [hunk 4]
```diff
@@ -88,21 +96,35 @@ export const ProgressStep = ({ ...props }) => {
...
-        <p className={`${prefix}--progress-optional`}>{secondaryLabel}</p>
-      ) : null}
-      <span className={`${prefix}--progress-line`} />
+      <div
+        className={classnames(`${prefix}--progress-step-button`, {
+          [`${prefix}--progress-step-button--unclickable`]: !onClick || current,
+        })}
+        role="button"
+        tabIndex={!current && onClick ? 0 : -1}
+        onClick={!current ? onClick : undefined}
```
Comment (emyarod) [inline src/components/ProgressIndicator/ProgressIndicator.js]: the PR description states that each progress step will have a click handler attached, but it looks like the current progress step will not be tabbable or receive the click handler. I just want to confirm, is this the expected behavior?

Comment (scottdickerson) [reply src/components/ProgressIndicator/ProgressIndicator.js]: Yeah @emyarod in our user testing we got feedback that we shouldn't allow them to click on the current step.

Comment (emyarod) [reply src/components/ProgressIndicator/ProgressIndicator.js]: gotcha, thanks for clarifying @scottdickerson !

File: src/components/ProgressIndicator/ProgressIndicator.js [hunk 5]
```diff
@@ -199,25 +231,37 @@ export class ProgressIndicator extends Component {
...
 
-  renderSteps = () =>
-    React.Children.map(this.props.children, (child, index) => {
+  renderSteps = () => {
+    const { onChange } = this.props;
+
+    return React.Children.map(this.props.children, (child, index) => {
+      // only setup click handlers if onChange event is passed
+      const onClick =
+        onChange && index > -1 ? () => onChange(index) : undefined;
```
Comment (asudoh) [inline src/components/ProgressIndicator/ProgressIndicator.js]: Do you anticipate `index` being negative in `React.Children.map()` callback?

Comment (scottdickerson) [reply src/components/ProgressIndicator/ProgressIndicator.js]: you're right, great catch I've removed that in the latest commit.

**Pull request merged**